File: src/Blizka.Api/Common/ApiError.cs
namespace Blizka.Api.Common;

/// <summary>Структурированный payload ошибки внутри <see cref="ApiErrorResponse"/>.</summary>
/// <param name="Code">Машиночитаемый код ошибки (например, <c>VALIDATION_ERROR</c>).</param>
/// <param name="Message">Локализованное, actionable-сообщение ("что делать"), резолвится на сервере.</param>
/// <param name="Details">Структурированный контекст ошибки (например, требуемая/доступная сумма) или <c>null</c>.</param>
/// <param name="Action">Подсказка клиенту для CTA (например, <c>TOP_UP_SPARKS</c>) или <c>null</c>.</param>
public sealed record ApiError(string Code, string Message, object? Details, string? Action);

/// <summary>Обёртка для каждого неуспешного ответа API.</summary>
/// <param name="Error">Детали ошибки.</param>
public sealed record ApiErrorResponse(ApiError Error)
{
    public static ApiErrorResponse From(string code, string message, object? details = null, string? action = null)
        => new(new ApiError(code, message, details, action));
}

File: frontend/src/shared/api/api-error.ts
/**
 * Единственный тип ошибки, который бросает `apiRequest`.
 * Бэкенд отвечает `{ error: { code, message, details, action } }`, где `message`
 * уже локализован сервером — его можно показывать пользователю как есть.
 */
export class ApiError extends Error {
  readonly status: number
  /** `VALIDATION_ERROR`, `INSUFFICIENT_SPARKS`, … */
  readonly code: string
  readonly details: unknown
  /** Подсказка для кнопки: `TOP_UP_SPARKS`, `COMPLETE_ONBOARDING`, … */
  readonly action: string | null

  constructor(params: {
    status: number
    code: string
    message: string
    details?: unknown
    action?: string | null
  }) {
    super(params.message)
    this.name = 'ApiError'
    this.status = params.status
    this.code = params.code
    this.details = params.details
    this.action = params.action ?? null
  }

  get isUnauthorized(): boolean {
    return this.status === 401
  }

  get isForbidden(): boolean {
    return this.status === 403
  }

  get isNotFound(): boolean {
    return this.status === 404
  }

  get isServer(): boolean {
    return this.status >= 500
  }
}

export function isApiError(error: unknown): error is ApiError {
  return error instanceof ApiError
}
